### PR TITLE
Library Forwarding: Add support for 32-bit Wayland

### PR DESCRIPTION
**Pull request by neobrain**
## Overview

Finally, the first real 32-bit library we can forward to the 64-bit host, moving us up to Phase 3 of the [Library Forwarding Roadmap](https://gist.github.com/neobrain/3812c111a2da57bfd8376dd21d61a886):
|                          | X11 (64-bit) | Wayland (64-bit) | Wayland (32-bit) | X11 (32-bit) |
| ------------------------ | ------------ | ---------------- | ---------------- |------------- |
| no thunks            | ☑️ 2022      | ☑️ 2022         | ☑️ 2022          | ☑️ 2022      |
| only Wayland thunks  | N/A          | ☑️ Phase 1      | ✅Phase 3          | N/A           |
| Vulkan               | ☑️ 2022      | ☑️ Phase 1      | Phase 4          | Phase 7       |
| zink on Vulkan       | ☑️ Phase 2   | ☑️ Phase 2      | Phase 4          | Phase 7       |
| OpenGL               | ☑️ 2022      | Phase 5         | Phase 6          | Phase 7       |

To enable this support, the `thunkgen` features added recently are coming together:
* **Opaque types**: `wl_proxy` is always created host-side and the guest only passes around the pointer. (Internally, we need to peek into implementation details to parse message signatures, but that doesn't concern `thunkgen`.)
* **Automatic struct repacking** (`wl_array`)
* **Customized struct repacking** (to handle the cyclic reference of `wl_interface` and`wl_message`, which even includes an array-of-pointers)
* **`ptr_passthrough`** and **`emit_layout_wrappers`** annotations (to send `guest_layout<wl_interface>` across architecture boundaries and trigger repacking manually, without compromising data layout compatibility checks elsewhere)

Additionally, a few new API functions are added to facilitate testing of this branch by making 32-bit Super Meat Boy and zink work.

## Testing

These configurations are confirmed to work when forwarding 32-bit Wayland:
* vkcube (via x86 Vulkan)
* Super Meat Boy (via x86 GL)
* Super Meat Boy (via x86 zink+Vulkan)

## Impact

The changes here are almost exclusively restricted to Wayland. Other than minor cleanups, the 64-bit wrapper should be unaffected.

## Future

Functioning Wayland forwarding paves the way for 32-bit Vulkan, which is much easier to test with working window system integration.

### CODE REVIEW

File: ThunkLibs/libwayland-client/Host.cpp [hunk 1]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+// Maps guest interface to host_interfaces
+static std::unordered_map<guest_layout<const wl_interface>*, wl_interface*> guest_to_host_interface;
+
+static wl_interface* get_proxy_interface(wl_proxy* proxy) {
+  // wl_proxy is a private struct, but its first member is the wl_interface pointer
+  return *reinterpret_cast<wl_interface**>(proxy);
+}
+
+static void assert_is_valid_host_interface(const wl_interface* interface) {
+  if ((uint32_t)interface->method_count >= 0x1000 || (uint32_t)interface->event_count >= 0x1000) {
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:57]: Should have a comment here explaining what's going on.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 2]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+  fprintf(stderr, "Unknown wayland interface %p, adding to registry\n", interface.get_pointer());
+
+  host_interface_it->second = new wl_interface;
+  wl_interface* host_interface = host_interface_it->second;
+  repack_guest_wl_interface_to_host(interface, host_interface);
+  return host_interface_it->second;
+}
+
+void fex_custom_repack_entry(host_layout<wl_interface>& into, guest_layout<wl_interface> const& from) {
+  // TODO: Manage lifetime of this memory?
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:106]: This isn't actually needed. `wl_interface`s correspond to global symbols on the guest side, so their lifetime is bound to program execution.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 3]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+  // TODO: Only copy as many arguments as needed by the message signature
+  std::array<wl_argument, WL_CLOSURE_MAX_ARGS> host_args;
+  for (int i = 0; i < host_args.size(); ++i) {
+    // NOTE: wl_argument can store a pointer argument, so for 32-bit guests
+    //       we need to make sure the upper 32-bits are explicitly zeroed
+    std::memset(&host_args[i], 0, sizeof(host_args[i]));
+    std::memcpy(&host_args[i], &args.get_pointer()[i], sizeof(args.get_pointer()[i]));
+  }
+  return host_args;
+}
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:170]: Should look into making this 32-bit-exclusive, since the extra copy isn't needed on 64-bit.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 4]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+    for (auto* ptr = method->signature; *ptr; ++ptr) {
+      if (isalpha(*ptr)) {
+        ++arg_count;
+      }
+    }
+
+    for (int arg_idx = 0; arg_idx < arg_count; ++arg_idx) {
+      if (method->types[arg_idx] != nullptr && args[arg_idx].o != nullptr) {
+        fprintf(stderr, "FOUND ARG %p OF TYPE %s\n", args[arg_idx].o, method->types[arg_idx]->name);
+        // TODO: Validate host_args[arg_idx].o is a valid 32-bit pointer
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:186]: Should address this TODO.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 5]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+fex_wl_proxy_marshal_array(
+            wl_proxy *proxy, uint32_t opcode,
+            guest_layout<wl_argument*> args,
+            guest_layout<const wl_interface*> guest_interface,
+            bool constructor, // Call the _constructor variant of the native wayland function
+            std::optional<uint32_t> version,
+            std::optional<uint32_t> flags) {
+  auto interface = lookup_wl_interface(guest_interface);
+
+  fprintf(stderr, "WAYLAND HOST: Marshaling %s with proxy %p, using interface %p (%s) method %d (max: %d)\n", __FUNCTION__, proxy, interface, interface ? interface->name : "(null)", opcode, interface ? interface->method_count : 0);
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:211]: Should drop these debug logging messages.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 6]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+
+  using PackedArgumentsType = PackedArguments<Result, guest_layout<Args>..., guest_layout<wl_array*>>;
+
+#ifndef IS_32BIT_THUNK
+  static_assert(has_compatible_data_layout<wl_array>);
+  guest_layout<wl_array*> guest_array_ptr = to_guest(to_host_layout(array));
+
+  auto packed_args = PackedArgumentsType {
+    to_guest(to_host_layout(args))..., guest_array_ptr
+  };
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:281]: Check if we can just use `CallbackUnpack::CallGuestPtr` directly here.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 7]
```diff
@@ -46,28 +315,38 @@ static void WaylandFinalizeHostTrampolineForGuestListener(void (*callback)()) {
 }
 
 extern "C" int fexfn_impl_libwayland_client_wl_proxy_add_listener(struct wl_proxy *proxy,
-      guest_layout<void (**)(void)> callback_raw, void* data) {
-  auto guest_interface = ((wl_proxy_private*)proxy)->interface;
+      guest_layout<void (**)(void)> callback_table_raw, void* data) {
+  // TODO: Rename to host_interface
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:319]: Should be addressed.

File: ThunkLibs/libwayland-client/libwayland-client_interface.cpp [hunk 8]
```diff
@@ -60,12 +75,33 @@ template<> struct fex_gen_config<wl_proxy_set_tag> {};
...
+template<> struct fex_gen_param<wl_proxy_marshal_array_flags, 5, wl_argument*> : fexgen::ptr_passthrough {};
 #endif
 
 // Guest notifies host about its interface. Host returns its corresponding interface pointer
-wl_interface* fex_wl_exchange_interface_pointer(wl_interface*, const char* name);
+void fex_wl_exchange_interface_pointer(wl_interface*, const char* name);
 template<> struct fex_gen_config<fex_wl_exchange_interface_pointer> : fexgen::custom_host_impl {};
+template<> struct fex_gen_param<fex_wl_exchange_interface_pointer, 0, wl_interface*> : fexgen::ptr_passthrough {};
+
+// This is equivalent to reading ((wl_proxy_private*)proxy)->interface->methods[opcode].signature on 64-bit.
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/libwayland-client_interface.cpp:98]: Drop `wl_proxy_private` mention, since it's no longer defined.

File: ThunkLibs/libwayland-client/Host.cpp [hunk 9]
```diff
@@ -21,13 +22,281 @@ tags: thunklibs|wayland-client
...
+  for (int i = 0; i < host_args.size(); ++i) {
+    // NOTE: wl_argument can store a pointer argument, so for 32-bit guests
+    //       we need to make sure the upper 32-bits are explicitly zeroed
+    std::memset(&host_args[i], 0, sizeof(host_args[i]));
+    std::memcpy(&host_args[i], &args.get_pointer()[i], sizeof(args.get_pointer()[i]));
+  }
+  return host_args;
+}
+
+static void fex_wl_validate_method_call(const wl_interface* interface, uint32_t opcode, std::span<wl_argument> args) {
```
Comment (neobrain) [inline ThunkLibs/libwayland-client/Host.cpp:172]: Needs `<span>`.

**Comment by neobrain** [comment]
While rebasing my WIP branch on this, I found some interesting edge cases in the interaction of `ptr_passthrough` and the fixed-size mapping of guest integer types. I refined the patch `Don't register types used exclusively with ptr_passthrough annotations` to support these edge cases, and I extended the corresponding unit tests to add coverage.

**Comment by neobrain** [comment]
Rebased on latest main.

**Review by Sonicadvance1 — APPROVED** [review]
